libxcrypt: Build only with strong hashes

mweinelt opened the pull request:
Effectively removes support for the following hashing algorithms as announced in the NixOS 22.11 release notes:

- bcrypt_x ($2x$)
- sha256crypt ($5$)
- sha1crypt ($sha1$)
- sunmd5 ($md5$)
- md5crypt ($1$)
- nt ($3$)
- bdiscrypt (_)
- bigcrypt (:)
- descrypt (:)

Passthru tests (login, shadow) built on aarch64-linux and x86_64-linux.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin (fails early during perl build)
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
libxcrypt: Build only with strong hashes

Effectively removes support for the following hashing algorithms
as announced in the NixOS 22.11 release notes:

- bcrypt_x ($2x$)
- sha256crypt ($5$)
- sha1crypt ($sha1$)
- sunmd5 ($md5$)
- md5crypt ($1$)
- nt ($3$)
- bdiscrypt (_)
- bigcrypt (:)
- descrypt (:)

And exposes the crypt scheme ids for enabled algorithms, so they can be
reused for validation in the users-groups module.

SuperSandro2000 commented:
Do we need to change https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/users-groups.nix#L705-L734 ?

mweinelt commented:
It does not validate the scheme so far, and I don't have the energy to make it do that.

mweinelt committed:
nixos/users-groups: Update password scheme validation

Updates the warnings message for statefully set up passwords, now that
weak algorithms have been removed from our libxcrypt package.

Additionall we now add proper validation for hashing schemes used in
`hashedPassword`.

Neither will prevent a rebuiild, but instead issue a warning, that this
requires immediate remediation, or else users will be unable to login.

Reuses the crypt scheme ids as provided by the libxcrypt package.

mweinelt commented:
Updated the validation for `hashedPassword`. Was quite a bit simpler than expected.

cc @rnhmjoj @SuperSandro2000

mweinelt commented on nixos/modules/config/users-groups.nix:
```diff
@@ -716,7 +716,8 @@ in {
           sep = "\\$";
           base64 = "[a-zA-Z0-9./]+";
-          id = "[a-z0-9-]+";
+          # must match the labels of hashes enabled in libxcrypt
+          id = "(y|gy|7|2b|2y|2a|2x|6)";
```
```
#name          h_prefix  nrbytes  flags
yescrypt       $y$       16       STRONG,DEFAULT,ALT,DEBIAN,FEDORA
gost_yescrypt  $gy$      16       STRONG,ALT
scrypt         $7$       16       STRONG
bcrypt         $2b$      16       STRONG,DEFAULT,ALT,FREEBSD,NETBSD,OPENBSD,OWL,SOLARIS,SUSE
bcrypt_y       $2y$      16       STRONG,ALT,OWL,SUSE
bcrypt_a       $2a$      16       STRONG,ALT,FREEBSD,NETBSD,OPENBSD,OWL,SOLARIS,SUSE
sha512crypt    $6$       15       STRONG,DEFAULT,GLIBC,FREEBSD,SOLARIS
```

https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L41

mweinelt commented on nixos/modules/config/users-groups.nix:
```diff
@@ -716,7 +716,8 @@ in {
           sep = "\\$";
           base64 = "[a-zA-Z0-9./]+";
-          id = "[a-z0-9-]+";
+          # must match the labels of hashes enabled in libxcrypt
+          id = "(y|gy|7|2b|2y|2a|6)";
```
```
#name          h_prefix  nrbytes  flags
yescrypt       $y$       16       STRONG,DEFAULT,ALT,DEBIAN,FEDORA
gost_yescrypt  $gy$      16       STRONG,ALT
scrypt         $7$       16       STRONG
bcrypt         $2b$      16       STRONG,DEFAULT,ALT,FREEBSD,NETBSD,OPENBSD,OWL,SOLARIS,SUSE
bcrypt_y       $2y$      16       STRONG,ALT,OWL,SUSE
bcrypt_a       $2a$      16       STRONG,ALT,FREEBSD,NETBSD,OPENBSD,OWL,SOLARIS,SUSE
sha512crypt    $6$       15       STRONG,DEFAULT,GLIBC,FREEBSD,SOLARIS
```

https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L41

mweinelt commented:
Tested with a sha256-crypt hash, and it does complain alright. Migated to yescrypt and the warning was gone.

```
trace: warning: The password hash of user "hexa" may be invalid. You must set a
valid hash or the user will be locked out of their account. Please
check the value of option `users.users."hexa".hashedPassword`.
```

mweinelt committed:
pam: Make libxcrypt a non-optional dependency

Our PAM configuration now defaults to yescrypt, which requires
libxcrypt.

mweinelt committed:
treewide: Make yescrypt the default algorithm for pam_unix.so

This ensures `passwd` will default to yescrypt for newly generated
passwords.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/21

rnhmjoj reviewed (commented):
The changes look good but I can't properly test this right now.
Anyway, glad to see we're finally using modern hashing schemes.

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -18,6 +18,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `nixos-rebuild` now supports an extra `--specialisation` option that can be used to change specialisation for `switch` and `test` commands.
 
+- `libxcrypt`, the library providing the `crypt(3)` password hashing function, will be built without support for algorithms not flagged [`strong`](https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L48). This affects the availability of password hashing algorithms used for the system login (`login(1)`, `passwd(1)`), but also Apache2 Basic-Auth, Samba, OpenLDAP, and [many other pac ...
```
```suggestion
- `libxcrypt`, the library providing the `crypt(3)` password hashing function, is now built without support for algorithms not flagged [`strong`](https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L48). This affects the availability of password hashing algorithms used for the system login (`login(1)`, `passwd(1)`), but also Apache2 Basic-Auth, Samba, OpenLDAP, and [many other packages](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20libxcrypt&type=code).
```

winterqt replied:
Agreed, but also:
```suggestion
- `libxcrypt`, the library providing the `crypt(3)` password hashing function, will be built without support for algorithms not flagged [`strong`](https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L48). This affects the availability of password hashing algorithms used for system login (`login(1)`, `passwd(1)`), but also Apache2 Basic-Auth, Samba, OpenLDAP, and [many other packages](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20libxcrypt&type=code).
```

mweinelt replied:
Applied both recommendations. Thanks.

fpletz reviewed (approved):
Not tested yet but the diff looks fine! :+1:

vcunat commented:
`libxcrypt-legacy` supporting all hashes:
https://github.com/NixOS/nixpkgs/pull/223034

vcunat commented:
- [ ] `mailutils` [fails tests](https://hydra.nixos.org/build/213354268/nixlog/4/tail).  Perhaps maintainers can have a look: @orivej @vrthra
- [ ] `lighttpd` [fails tests](https://hydra.nixos.org/build/213382643/nixlog/4/tail).  Perhaps maintainers can have a look: @bjornfor @brechtcs
- [ ] `radicale2` [fails tests](https://hydra.nixos.org/build/213373432/nixlog/3/tail). Perhaps maintainers can have a look: @edwtjo @pSub @infinisil

Generally I can imagine options around updating, fixing or disabling tests and maybe even using `libxcrypt-legacy`.

orivej commented:
I see that `mailutils` uses weak auth during tests to interact with its mock server, i.e. it sends plaintext `USER name`, `PASS  guessme` before sending other testing queries. Maybe we should ask its developers on the best way forward (https://savannah.gnu.org/support/?group=mailutils) and meanwhile use `libxcrypt-legacy` or disable tests.

obadz commented:
I continue to think this should be configurable, see #208603
